test_plans/SYCL-2020/aspect_selector.md

```diff
@@ -0,0 +1,102 @@
+Each test will construct an aspect selector object,
+which will try to select another device,
+this one based on the requested aspects (i.e. `selectedDevice`).
+`selectedDevice` can be any device in the system,
+not necessarily the same as `testDevice`.
+
+### Function overloads
+
+Each test that tests aspects `{asp1, ..., aspN}` must test all three overloads of the `aspect_selector` function:
+
+* Pass a vector of aspects: `aspect_selector(std::vector{asp1, ..., aspN})`
+* Pass aspects as function arguments: `aspect_selector(asp1, ..., aspN)`
+* Pass aspects as template parameters: `aspect_selector<asp1, ..., aspN>()`
+
+### Combinations of aspects
+
+All aspects listed in the `enum class aspect` will be tested,
+let's identify these aspects as `{asp1, ..., aspN}`.
+
+The CTS will cover these combinations:
+
+1. All combinations where no aspect is passed in.
+2. All combinations where a single aspect is used.
+3. All combinations of two aspects,
+including duplicates (e.g. `aspect_selector(aspX, aspX)`),
+and reversed order (e.g. `aspect_selector(aspX, aspY)` and `aspect_selector(aspY, aspX)`).
+
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`,
+where the aspects are not duplicated,
+i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
+
+> Note: it is recommended a Python generator is written to cover all required combinations.
+
+## Tests
+
+### No aspects
+
+1. `auto selector1 = aspect_selector()`
+2. `auto selector2 = aspect_selector<>()`
+
+In the test, both `selector1` and `selector2` are used
+to construct a SYCL device, `dev1` and `dev2` respectively.
+Both `dev1` and `dev2` must compare equal to the device obtained by `default_selector`
+(let's call this device `defaultDevice`).
+
+### Requested aspects using a device
+
+Assume requested aspects `{asp1, ..., aspN}` for each test.
+
+These steps are performed for each combination:
+
+1. Construct `testDevice`, which represents the device where conformance will be passed on.
+
+    * Also retrieve `testPlatform` SYCL platform from `testDevice` and construct `testQueue` SYCL queue from the device.
+2. Test using a vector of aspects
+
+    1. Construct `selectorVector` object obtained by calling `aspect_selector(std::vector{asp1, ..., aspN})`.
+    2. Try/catch block
+
+        1. Construct SYCL device (`selectedDevice`) using `selectorVector`.
+        2. If exception is throw:
+
+            * If `testDevice` contains all the requested aspects,
+              but no device was found using the aspect selector,
+              this is a failed test.
+            * Otherwise there is most likely no device in the system with the requested aspects, skip test.
```

Comment:
Why not iterate over all the devices in the system and see if there is one with all the requested aspects?

Reply:
Changed it to that, separated the platform.